src/allmydata/storage/server.py

```diff
@@ -340,6 +354,21 @@ def remote_allocate_buckets(self, storage_index,
         self.add_latency("allocate", self._get_current_time() - start)
         return alreadygot, bucketwriters
 
+    def remote_allocate_buckets(self, storage_index,
+                                renew_secret, cancel_secret,
+                                sharenums, allocated_size,
+                                canary, owner_num=0):
+        """Foolscap-specific ``allocate_buckets()`` API."""
+        alreadygot, bucketwriters = self._allocate_buckets(
+            storage_index, renew_secret, cancel_secret, sharenums, allocated_size,
+            owner_num=owner_num,
+        )
+        # Abort BucketWriters if disconnection happens.
+        for bw in bucketwriters.values():
+            disconnect_marker = canary.notifyOnDisconnect(bw.disconnected)
+            self._bucket_writer_disconnect_markers[bw] = (canary, disconnect_marker)
+        return alreadygot, bucketwriters
```

Comment:
nice, and makes me wonder if a further refactoring will make sense at some point - to put the Foolscap parts in a separate class from the generic storage logic parts.

Reply:
Quite possibly. As first pass probably there will be `remote_X()` vs `X()` methods, but it might turn into wrapper I suppose.